File: LeetCode/dp/198.打家劫舍.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        f1 = f0 = 0
        for i, x in enumerate(nums):
            new_f = max(f1, f0 + x)
            f0 = f1
            f1= new_f
        return f1

File: LeetCode/dp/198.打家劫舍.cpp
//
// Created by lbyang on 23-10-25.
//
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int dp[150][2];
        dp[0][0] = nums[0];
        dp[0][1] = 0;
        for(int i = 1;i < nums.size(); i++) {
            dp[i][0] = max(dp[i - 1][1] + nums[i], dp[i - 1][0]);
            dp[i][1] = max(dp[i-1][0], dp[i-1][1]);
        }
        return max(dp[nums.size()-1][0], dp[nums.size()-1][1]);
    }
};